### PR TITLE
Transaction Watcher initial interface

### PR DESCRIPTION
**Pull request by avonian**
## Description

Looking to build a TransactionWatcher (inspired by OrderWatcher) which could be used as a higher-level API for watching events. This can be especially useful for handling failed transactions (e.g. out of gas).

## Motivation and Context

The current recommended approach for detecting failed transactions is to poll the blockchain for transaction receipts. After discussing the problem with 0x team members, it seems this would be a good use case for a TransactionWatcher (similar to OrderWatcher).

## How Has This Been Tested?

Work in progress.

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [x] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [ ] Change requires a change to the documentation.
*   [ ] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOGs.
*   [ ] Updated the new versions of the changed packages in the relevant CHANGELOGs.
*   [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/0x.js/src/order_watcher/transaction_watcher.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    public addTransaction(transactionHash: string): void {
+        assert.isHexString('transactionHash', transactionHash);
+        this._transactionHashes.push(transactionHash)
+    }
+    /**
+     * Removes a transaction from the transactionWatcher
+     * @param   transactionHash     The transactionHash of the transaction you wish to stop watching.
+     */
+    public removeTransaction(orderHash: string): void {
```
Comment (LogvinovLeon) [inline packages/0x.js/src/order_watcher/transaction_watcher.ts]: Do we need that? Transaction can be auto-removed when it's settled or not settled for N blocks.

File: packages/0x.js/src/types.ts [hunk 2]
```diff
@@ -283,4 +285,8 @@ export interface OrderStateInvalid {
 export type OrderState = OrderStateValid | OrderStateInvalid;
 
 export type OnOrderStateChangeCallback = (err: Error | null, orderState?: OrderState) => void;
+
+export type OnTransactionSuccessCallback = (transactionReceipt?: TransactionReceiptWithDecodedLogs) => void;
```
Comment (LogvinovLeon) [inline packages/0x.js/src/types.ts]: Why is that parameter optional?

File: packages/0x.js/src/types.ts [hunk 3]
```diff
@@ -283,4 +285,8 @@ export interface OrderStateInvalid {
...
+
+export type OnTransactionFailedCallback = (transactionReceipt?: TransactionReceipt) => void;
```
Comment (LogvinovLeon) [inline packages/0x.js/src/types.ts]: And this one too

**Comment by avonian** [comment]
I wanted to submit something quickly to at least get the ball rolling, look forward to everyone's feedback.

File: packages/0x.js/src/order_watcher/transaction_watcher.ts (hunk 1 above)
Comment (avonian) [reply packages/0x.js/src/order_watcher/transaction_watcher.ts]: Good point, will remove.

File: packages/0x.js/src/types.ts (hunk 2 above)
Comment (avonian) [reply packages/0x.js/src/types.ts]: Typical mistake copy and pasting, noted for fix.

File: packages/0x.js/src/types.ts (hunk 3 above)
Comment (avonian) [reply packages/0x.js/src/types.ts]: Typical mistake copy and pasting, noted for fix.

**Comment by LogvinovLeon** [comment]
Reopen if still valid

**Pull request closed without merging**